lib/datadog/appsec/contrib/rack/request_middleware.rb

```diff
@@ -49,6 +54,50 @@ def call(env)
             AppSec::Event.record(*both_response.map { |_action, event| event }) if both_response.any?
 
             request_return
+          ensure
+            add_waf_runtime_tags(context)
+            request!
+          end
+
+          private
+
+          def first_request?
+            @first_request
+          end
+
+          def request!
+            @first_request = false
+          end
+
+          def active_trace
+            return unless defined?(Datadog::Tracing) && Datadog::Tracing.respond_to?(:active_span)
+
+            Datadog::Tracing.active_trace
+          end
+
+          def add_appsec_tags
+            return unless active_trace
```

Comment:
Minor: There's a potential here for a weird interaction. If the customers start with tracing disabled (or for some reason have a special request that skips tracing), we'll mark the request as first, but never actually add this information (due to the `return unless active_trace`).

Should we perhaps only set that flag _after_ sending it on a trace?

Reply:
As it stands, if the app starts with `Tracing` disabled there is no trace that would ever be sent, therefore it becomes immaterial. So AppSec would presumably work but be unable to send any data to the backend. That is, until traces become a proper `Core` data bus independent of `Tracing`.

That might change with remote configuration so I understand the logic here.

> for some reason have a special request that skips tracing

Is that possible?

Still, I will see if I can come up with something.

Reply:
> As it stands, if the app starts with Tracing disabled there is no trace that would ever be sent, therefore it becomes immaterial. So AppSec would presumably work but be unable to send any data to the backend. That is, until traces become a proper Core data bus independent of Tracing.

Right. My point is that in such a case, we'll still mark the request as not being the first one, and in such a case the special "first request" metadata would never be sent.

> > for some reason have a special request that skips tracing
>
> Is that possible?

I guess it could if customers enable appsec and its instrumentation, then do a request, and then enable tracing and its instrumentation.

Probably both of the cases I'm pointing out here are quite corner cases, but since the solution is quite simple -- don't mark a request as first until we actually send the special "first request" metadata I think it may be worth fixing.

(The change I propose above probably makes `@first_request` a bit misnamed, so probably it would become `@appsec_info_sent_once` or something like that?)

Reply:
Refactored into something like that.